src/Psl/Option/Option.php

```diff
@@ -0,0 +1,260 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Option;
+
+use Closure;
+
+/**
+ * @template T
+ */
+final class Option
+{
+    /**
+     * @param ?array{T} $option
+     *
+     * @internal
+     */
+    private function __construct(
+        private readonly null|array $option,
+    ) {
+    }
+
+    /**
+     * Create an option with some value.
+     *
+     * @template Tv
+     *
+     * @param Tv $value
+     *
+     * @return Option<Tv>
+     */
+    public static function some(mixed $value): Option
+    {
+        return new self([$value]);
+    }
+
+    /**
+     * Create an option with none value.
+     *
+     * @template Tn
+     *
+     * @return Option<Tn>
+     */
+    public static function none(): Option
+    {
+        return new self(null);
+    }
+
+    /**
+     * Returns true if the option is a some value.
+     */
+    public function isSome(): bool
+    {
+        return $this->option !== null;
+    }
+
+    /**
+     * Returns true if the option is a some and the value inside of it matches a predicate.
+     *
+     * @param (Closure(T): bool) $predicate
+     */
+    public function isSomeAnd(Closure $predicate): bool
+    {
+        return $this->option !== null && $predicate($this->option[0]);
+    }
+
+    /**
+     * Returns true if the option is a none.
+     */
+    public function isNone(): bool
+    {
+        return $this->option === null;
+    }
+
+    /**
+     * Returns the contained Some value, consuming the self value.
+     *
+     * because this function may throw, its use is generally discouraged.
+     * Instead, prefer to use `Option::unwrapOr()`, `Option::unwrapOrElse()`.
+     *
+     * @throws Exception\NoneException If the option is none.
+     *
+     * @return T
+     */
+    public function unwrap(): mixed
+    {
+        if ($this->option !== null) {
+            return $this->option[0];
+        }
+
+        throw new Exception\NoneException('Attempting to unwrap a none option.');
+    }
+
+    /**
+     * Returns the contained some value or the provided default.
+     *
+     * @note:   Arguments passed to `Option::unwrapOr()` are eagerly evaluated;
+     *          if you are passing the result of a function call, it is recommended to use `Option::unwrapOrElse()`, which is lazily evaluated.
+     *
+     * @param T $default
+     *
+     * @return T
+     */
+    public function unwrapOr(mixed $default): mixed
```

Comment:
you have both `unwrapOr` and `unwrapOrElse` which are basically the same, but accept another input argument type. This might be a bit confusing. Can we solve this by playing around with the name?

Maybe `unwrapOrFallback` and `unwrapOrCalculate`?

Reply:
@veewee `unwrapOr()` accepts already computed default value, while `unwrapOrThen()` accepts a **callback computing the default value**, which means the former is non-lazy, and the latter is lazy, which might be desirable. Also, a callback may throw, so you can do it like

```php
$option->unwrapOrElse(fn() => throw new \LogicException('Unexpected None'));
```

Reply:
generally I'd avoid too long names like `unwrapOrFallback` - these functions are expected to be used pretty often, so it's okay to shorten their names.